File: packages/backend/src/middleware/rateLimit.ts
/**
 * 限流中间件
 * 使用 Redis 实现滑动窗口限流
 */

import { Request, Response, NextFunction } from 'express'
import { createLogger } from '../utils/logger.js'
import { getRedisClient } from '../storage/redis.js'
import { getConfig } from '../config/index.js'

const logger = createLogger('RateLimitMiddleware')

const RATE_LIMIT_PREFIX = 'ratelimit:'

/**
 * 限流中间件
 */
export async function rateLimitMiddleware(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  const config = getConfig()

  // 如果未启用限流，直接放行
  if (!config.rateLimit.enabled) {
    next()
    return
  }

  // 健康检查端点不限流
  if (req.path === '/health' || req.path === '/api/admin/health') {
    next()
    return
  }

  const redis = getRedisClient()
  const windowMs = config.rateLimit.windowMs
  const maxRequests = config.rateLimit.maxRequests

  // 使用 IP 作为限流 key
  const clientId = req.ip || req.socket.remoteAddress || 'unknown'
  const key = RATE_LIMIT_PREFIX + clientId

  try {
    const now = Date.now()
    const windowStart = now - windowMs

    // 使用 Redis 事务实现滑动窗口
    const pipeline = redis.pipeline()

    // 移除窗口外的请求记录
    pipeline.zremrangebyscore(key, 0, windowStart)

    // 获取当前窗口内的请求数
    pipeline.zcard(key)

    // 添加当前请求
    pipeline.zadd(key, now, `${now}-${Math.random()}`)

    // 设置过期时间
    pipeline.expire(key, Math.ceil(windowMs / 1000))

    const results = await pipeline.exec()

    if (results) {
      const currentCount = results[1][1] as number

      // 设置响应头
      res.setHeader('X-RateLimit-Limit', maxRequests)
      res.setHeader('X-RateLimit-Remaining', Math.max(0, maxRequests - currentCount - 1))
      res.setHeader('X-RateLimit-Reset', Math.ceil((now + windowMs) / 1000))

      if (currentCount >= maxRequests) {
        logger.warn('Rate limit exceeded', { clientId, count: currentCount })

        res.status(429).json({
          error: {
            message: 'Too many requests, please try again later',
            type: 'rate_limit_error'
          }
        })
        return
      }
    }

    next()
  } catch (error) {
    // 限流失败时放行，避免影响正常请求
    logger.error('Rate limit check failed', { error: (error as Error).message })
    next()
  }
}

/**
 * 创建自定义限流中间件
 */
export function createRateLimiter(options: {
  windowMs: number
  maxRequests: number
  keyPrefix?: string
}) {
  return async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    const redis = getRedisClient()
    const { windowMs, maxRequests, keyPrefix = 'custom:' } = options

    const clientId = req.ip || req.socket.remoteAddress || 'unknown'
    const key = RATE_LIMIT_PREFIX + keyPrefix + clientId

    try {
      const now = Date.now()
      const windowStart = now - windowMs

      const pipeline = redis.pipeline()
      pipeline.zremrangebyscore(key, 0, windowStart)
      pipeline.zcard(key)
      pipeline.zadd(key, now, `${now}-${Math.random()}`)
      pipeline.expire(key, Math.ceil(windowMs / 1000))

      const results = await pipeline.exec()

      if (results) {
        const currentCount = results[1][1] as number

        res.setHeader('X-RateLimit-Limit', maxRequests)
        res.setHeader('X-RateLimit-Remaining', Math.max(0, maxRequests - currentCount - 1))

        if (currentCount >= maxRequests) {
          res.status(429).json({
            error: {
              message: 'Too many requests',
              type: 'rate_limit_error'
            }
          })
          return
        }
      }

      next()
    } catch (error) {
      logger.error('Custom rate limit check failed', { error: (error as Error).message })
      next()
    }
  }
}

File: packages/gateway/middleware/rate_limit.go
package middleware

import (
	"context"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/rs/zerolog/log"

	"github.com/kiro-gateway/gateway/config"
	"github.com/kiro-gateway/gateway/storage"
	"github.com/kiro-gateway/gateway/util"
)

const rateLimitPrefix = "ratelimit:"

// RateLimitMiddleware returns a Gin middleware that enforces per-IP rate
// limiting using a Redis sorted-set sliding window. If rate limiting is
// disabled in config, the middleware is a no-op. On Redis errors the request
// is allowed through (graceful degradation).
func RateLimitMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		cfg := config.GetConfig()
		if !cfg.RateLimit.Enabled {
			c.Next()
			return
		}

		clientIP := c.ClientIP()
		key := storage.Key(rateLimitPrefix + clientIP)
		windowMs := int64(cfg.RateLimit.WindowMs)
		maxRequests := cfg.RateLimit.Max
		now := time.Now().UnixMilli()
		windowStart := now - windowMs
		ctx := context.Background()
		rdb := storage.GetClient()

		// Execute sliding window commands in a pipeline.
		pipe := rdb.Pipeline()

		// Remove entries outside the current window.
		pipe.ZRemRangeByScore(ctx, key, "-inf", strconv.FormatInt(windowStart, 10))

		// Add the current request timestamp.
		pipe.ZAdd(ctx, key, newZMember(float64(now), strconv.FormatInt(now, 10)))

		// Count entries in the window.
		countCmd := pipe.ZCard(ctx, key)

		// Set key expiry to the window duration.
		expiry := time.Duration(windowMs) * time.Millisecond
		pipe.Expire(ctx, key, expiry)

		if _, err := pipe.Exec(ctx); err != nil {
			// Graceful degradation: allow request on Redis error.
			log.Warn().Err(err).Str("ip", clientIP).Msg("Rate limit Redis error, allowing request")
			c.Next()
			return
		}

		count := countCmd.Val()
		remaining := int64(maxRequests) - count
		if remaining < 0 {
			remaining = 0
		}
		resetTime := now + windowMs

		// Set rate limit response headers.
		c.Header("X-RateLimit-Limit", strconv.Itoa(maxRequests))
		c.Header("X-RateLimit-Remaining", strconv.FormatInt(remaining, 10))
		c.Header("X-RateLimit-Reset", strconv.FormatInt(resetTime, 10))

		if count > int64(maxRequests) {
			util.ErrorJSON(c, http.StatusTooManyRequests,
				fmt.Sprintf("Rate limit exceeded. Try again in %d seconds", windowMs/1000),
				util.ErrTypeRateLimit)
			c.Abort()
			return
		}

		c.Next()
	}
}
